File: soungmunkim/Python/LinkedList/19_RemoveNthNodeFromEnd.py
"""
[[뒤에서 n번째 노드 지우기]]
Given the head of a linked list, 
remove the nth node from the end of the list and return its head.

Input: head = [1,2,3,4,5], n = 2
Output: [1,2,3,5]

Example 2:
Input: head = [1], n = 1
Output: []

Example 3:
Input: head = [1,2], n = 1
Output: [1]
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
# 리스트로부터 연결 리스트를 생성한 후 헤드 노드를 반환
def create_linked_list(l: list):
    if len(l) == 0:         # 리스트의 길이가 0인 경우
        return None
    
    start = ListNode(l[0])  # 시작 노드 생성
    node = start            # 현재 노드를 시작 노드로 설정
    
    for i in range(1, len(l)): # 리스트의 나머지 원소들을 순회
        node_new = ListNode(l[i])  # 새 노드 생성
        node.next = node_new      # 현재 노드의 다음을 새 노드로 연결
        node = node_new           # 현재 노드를 새 노드로 갱신
        
    return start    # 연결 리스트의 시작 노드 반환

# 연결 리스트의 모든 값을 출력
def print_linked_list(n: ListNode, l:list):
    if n is not None:    # 노드가 None이 아닌 경우
        l.append(n.val)  # 노드의 값을 리스트에 추가
        print_linked_list(n.next, l) # 다음 노드로 재귀 호출
    else:
        print(l)         # 노드가 None인 경우 리스트 출력
        

def removeNthFromEnd(head: ListNode, n: int) -> ListNode:
    dummy = ListNode(0)
    dummy.next = head

    front = back = dummy

    for i in range(n): #우선 뒷 포인터를 n개 먼저 움직인다. 
        back = back.next

    while back.next: #n개 떨어진 앞, 뒤 포인터가 끝까지 순차적으로 간다
        front = front.next
        back = back.next

    front.next = front.next.next #맨 마지막에 다다르면 앞포인터를 뒷포인터와 연결한다. 

    return dummy.next

def RemoveNodeFromEnd(head: list, n: int) -> ListNode:
    node = create_linked_list(head)
    output = removeNthFromEnd(node, n)
    return print_linked_list(output, [])


if __name__ == "__main__":
    # head = [1,2,3,4,5]
    # n = 2
    # RemoveNodeFromEnd(head, n)

    head = [1]
    n = 1
    RemoveNodeFromEnd(head, n)

File: soungmunkim/C++/LinkedList/19_RemoveNthNodeFromEnd.cpp
/*
[뒤에서 n번째 노드 지우기]
Given the head of a linked list, 
remove the nth node from the end of the list and return its head.

(base) soungmunkim@gimseongmun-ui-MacBookPro LinkedList % g++ -std=c++17 19_RemoveNthNodeFromEnd.cpp -o test
(base) soungmunkim@gimseongmun-ui-MacBookPro LinkedList % ./test
Enter the first list: [1,2,3,4,5]
Enter the integer(nth) that you want to remove: 2
[1, 2, 3, 5]

(base) soungmunkim@gimseongmun-ui-MacBookPro LinkedList % ./test
Enter the first list: [1]
Enter the integer(nth) that you want to remove: 1
[]

(base) soungmunkim@gimseongmun-ui-MacBookPro LinkedList % ./test
Enter the first list: [1,2]
Enter the integer(nth) that you want to remove: 1
[1]
*/

#include <iostream>
#include <vector>
#include <sstream>

using namespace std;

class ListNode {
public:
    int val;                  // 노드의 값
    ListNode *next;           // 다음 노드를 가리키는 포인터

    ListNode(int x) : val(x), next(NULL) {}  // 생성자

    // 주어진 숫자 배열을 바탕으로 연결 리스트를 생성하는 스태틱 함수
    static ListNode* create(const vector<int>& nums) {
        if (nums.empty()) {
            return nullptr;  // 배열이 비어있다면 nullptr 반환
        }

        ListNode* start = new ListNode(nums[0]);
        ListNode* node = start;

        for (int i = 1; i < nums.size(); i++) {
            ListNode* node_new = new ListNode(nums[i]);
            node->next = node_new;
            node = node_new;
        }

        return start;  // 생성된 연결 리스트의 첫 번째 노드 반환
    }

    // 연결 리스트의 모든 노드의 값을 출력하는 함수
    void print() {
        cout << "[";
        ListNode* node = this;
        while (node->next) {
            cout << node->val << ", ";
            node = node->next;
        }
        cout << node->val << "]" << endl;
    }
};

ListNode* removeNthFromEnd(ListNode* head, int n) {
    // 더미 노드를 생성하고 초기화한다.
    ListNode* dummy = new ListNode(0);
    dummy->next = head;

    ListNode* front = dummy;
    ListNode* back = dummy;

    // 뒷 포인터를 n개 먼저 움직인다.
    for (int i = 0; i < n; i++) {
        back = back->next;
    }

    // n개 떨어진 앞, 뒤 포인터가 끝까지 순차적으로 간다.
    while (back->next) {
        front = front->next;
        back = back->next;
    }

    // 맨 마지막에 다다르면 앞 포인터의 다음 노드를 뒷 포인터와 연결한다.
    front->next = front->next->next;

    // dummy 노드의 다음 노드를 반환한다.
    ListNode* result = dummy->next;
    delete dummy;
    return result;
}

int main() {
    // 첫 번째 연결 리스트 입력 받기
    cout << "Enter the first list: ";
    string line;
    getline(cin, line);

    // 불필요한 문자 제거
    for (auto& c : line) {
        if (c == '[' || c == ']' || c == ',') {
            c = ' ';
        }
    }

    istringstream iss1(line);
    int num;
    vector<int> v;
    while (iss1 >> num) {
        v.push_back(num);
    }
    ListNode* head = ListNode::create(v);

    // 지우고 싶은 뒤에서 n번째 받기
    cout << "Enter the integer(nth) that you want to remove: ";
    int n;
    cin >> n;

    // 뒤에서 nth node 지우기
    ListNode* result = removeNthFromEnd(head, n);
    if (result) {
        result->print();  // 결과 출력
    } else {
        cout << "[]\n";   // 빈 연결 리스트를 나타냄
    }


    // 메모리 해제를 위한 코드는 여기에 추가 구현이 필요합니다.
    return 0;
}